integrations/object_store/src/service/mod.rs

```diff
@@ -0,0 +1,380 @@
+
+use crate::utils::format_metadata;
+
+/// ObjectStore backend builder
+#[derive(Default)]
+pub struct ObjectStoreBuilder {
+    store: Option<Arc<dyn ObjectStore + 'static>>,
+}
+
+impl Debug for ObjectStoreBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut d = f.debug_struct("ObjectStoreBuilder");
+        d.finish_non_exhaustive()
+    }
+}
+
+impl ObjectStoreBuilder {
+    /// Set the object store instance
+    pub fn new(store: Arc<dyn ObjectStore + 'static>) -> Self {
+        Self { store: Some(store) }
+    }
+}
+
+impl Builder for ObjectStoreBuilder {
+    type Config = ();
+
+    fn build(self) -> Result<impl Access> {
+        let store = self.store.ok_or_else(|| {
+            Error::new(ErrorKind::ConfigInvalid, "object store is required")
+                .with_context("service", Scheme::Custom("object_store"))
+        })?;
+
+        Ok(ObjectStoreService { store })
+    }
+}
+
+/// ObjectStore backend
+pub struct ObjectStoreService {
+    store: Arc<dyn ObjectStore + 'static>,
+}
+
+impl Debug for ObjectStoreService {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut d = f.debug_struct("ObjectStoreBackend");
+        d.finish_non_exhaustive()
+    }
+}
+
+impl Access for ObjectStoreService {
+    type Reader = ObjectStoreReader;
+    type Writer = MultipartWriter<ObjectStoreWriter>;
+    type Lister = ObjectStoreLister;
+    type Deleter = BatchDeleter<ObjectStoreDeleter>;
+
+    fn info(&self) -> Arc<AccessorInfo> {
+        let info = AccessorInfo::default();
+        info.set_scheme("object_store")
+            .set_root("/")
+            .set_name("object_store")
+            .set_native_capability(Capability {
+                stat: true,
+                stat_with_if_match: true,
+                stat_with_if_unmodified_since: true,
+                read: true,
+                write: true,
+                delete: true,
+                list: true,
+                list_with_limit: true,
+                list_with_start_after: true,
+                delete_with_version: false,
+                ..Default::default()
+            });
+        Arc::new(info)
+    }
+
+    async fn stat(&self, path: &str, _: OpStat) -> Result<RpStat> {
+        let path = object_store::path::Path::from(path);
+        let meta = self.store.head(&path).await.map_err(parse_error)?;
+        let metadata = format_metadata(&meta);
+        Ok(RpStat::new(metadata))
+    }
+
+    async fn read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::Reader)> {
+        let reader = ObjectStoreReader::new(self.store.clone(), path, args).await?;
+        Ok((reader.rp(), reader))
+    }
+
+    async fn write(&self, path: &str, args: OpWrite) -> Result<(RpWrite, Self::Writer)> {
+        let writer = ObjectStoreWriter::new(self.store.clone(), path, args);
+        Ok((
+            RpWrite::default(),
+            MultipartWriter::new(self.info(), writer, 10),
```

Comment:
We could make concurrent configurable; that might be a nice improvement for the next PR.